### PR TITLE
Connect message queue

### PR DESCRIPTION
**Pull request by xitronix**

### CODE REVIEW

File: universal-login-relayer/lib/services/messages/MessageQueueService.ts [hunk 1]
```diff
@@ -21,20 +20,14 @@ class QueuedMessageService {
 
   async execute(signedMessage: SignedMessage, id: string) {
     try {
-      const transaction: providers.TransactionRequest = messageToTransaction(signedMessage);
-      const sentTransaction = await this.wallet.sendTransaction(transaction);
-      await this.provider.waitForTransaction(sentTransaction.hash!);
-      await this.onTransactionSent!(sentTransaction);
+      const sentTransaction = await this.messageExecutor.execute(signedMessage);
+      await this.messageExecutor.wallet.provider.waitForTransaction(sentTransaction.hash!);
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/lib/services/messages/MessageQueueService.ts]: move to Executor e.g. `executeAndWait`

File: universal-login-relayer/lib/services/messages/MessageQueueStore.ts [hunk 2]
```diff
@@ -39,13 +39,25 @@ export default class MessageQueueStore implements IMessageQueueStore {
   }
 
   async onSuccessRemove (id: string, hash: string) {
+    const messageField = await this.get(id);
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/lib/services/messages/MessageQueueStore.ts]: `async onSuccessRemove (id: string, hash: string)`  ->
`async markAsSuccess (messageId: string, transactionhHash: string)`  ->

Comment (JustynaBroniszewska) [reply universal-login-relayer/lib/services/messages/MessageQueueStore.ts]: `messageField` -> `messageEntity`

File: universal-login-relayer/lib/services/messages/MessageQueueStore.ts [hunk 3]
```diff
@@ -39,13 +39,25 @@ export default class MessageQueueStore implements IMessageQueueStore {
...
+    const messageHash = await calculateMessageHash(bignumberifySignedMessageFields(messageField.message));
+    await this.database('messages')
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/lib/services/messages/MessageQueueStore.ts]: That change should be triggered by `MessageQueueService` and executed by `PendingMessagesStore`

File: universal-login-relayer/lib/services/messages/MessageExecutor.ts [hunk 4]
```diff
@@ -6,7 +6,7 @@ import MessageValidator from './MessageValidator';
 
 export class MessageExecutor {
 
-  constructor(private wallet: Wallet, private onTransactionSent: OnTransactionSent, private messageValidator: MessageValidator) {
+  constructor(public wallet: Wallet, private onTransactionSent: OnTransactionSent, private messageValidator: MessageValidator) {
```
Comment (xitronix) [inline universal-login-relayer/lib/services/messages/MessageExecutor.ts]: Could be private

File: universal-login-commons/lib/utils/changingMessageFields.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import {utils} from 'ethers';
-import {SignedMessage} from '@universal-login/commons';
+import {SignedMessage} from '../.';
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/utils/changingMessageFields.ts]: Typo?

Comment (xitronix) [reply universal-login-commons/lib/utils/changingMessageFields.ts]: File was moved to commons. It works, but i can change this for '../'

File: universal-login-relayer/lib/routes/wallet.ts [hunk 6]
```diff
@@ -27,10 +27,26 @@ messageHandler : MessageHandler) => async (req : Request, res : Response, next :
...
   try {
     const {hash} = req.params;
-    const status = await transactionContractService.getStatus(hash);
+    const status = await messageHandler.getStatus(hash);
+    res.status(200)
+      .type('json')
+      .send(JSON.stringify(status));
+  } catch (err) {
+    next(err);
+  }
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/routes/wallet.ts]: #nextPR 
to my best knowlage  `try { ...} catch (err) {next(err);}` is only needed for asynchronious code, so we can skip it there

https://expressjs.com/en/guide/error-handling.html

File: universal-login-relayer/lib/routes/wallet.ts [hunk 7]
```diff
@@ -27,10 +27,26 @@ messageHandler : MessageHandler) => async (req : Request, res : Response, next :
...
+};
+
+export const getStatusById = (messageHandler: MessageHandler) => async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const {id} = req.params;
+    const messageEntity = await messageHandler.getQueueStatus(id);
+    const status = messageEntity && {
+      transactionHash: messageEntity.hash,
+      error: messageEntity.error
+    };
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/routes/wallet.ts]: Why not simply:
```ts
res.status(200)
    .type('json')
    .send(messageEntity);
```

File: universal-login-relayer/lib/services/messages/MessageQueueStore.ts [hunk 8]
```diff
@@ -53,7 +56,7 @@ export default class MessageQueueStore implements IMessageQueueStore {
       .first();
   }
 
-  private remove(id: string, data: QueueItem) {
+  private markAs(id: string, data: QueueItem) {
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/messages/MessageQueueStore.ts]: `update`

File: universal-login-relayer/lib/services/messages/PendingMessages.ts [hunk 9]
```diff
@@ -1,22 +1,21 @@
...
-  constructor(private wallet : Wallet, private messagesStore: IPendingMessagesStore, public doExecute: DoExecute) {
+  constructor(private wallet : Wallet, private messagesStore: IPendingMessagesStore, public messageQueue: MessageQueueService) {
   }
 
   async isPresent(messageHash : string) {
     return this.messagesStore.isPresent(messageHash);
   }
 
-  async add(message: SignedMessage) : Promise<providers.TransactionResponse | MessageStatus> {
+  async add(message: SignedMessage) : Promise<string | MessageStatus> {
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/messages/PendingMessages.ts]: Should be `Promise<MessageStatus>`, why two different types?

File: universal-login-relayer/test/config/message.ts [hunk 10]
```diff
@@ -1,7 +1,5 @@
-import {TEST_ACCOUNT_ADDRESS, createSignedMessage, Message} from '@universal-login/commons';
+import {TEST_ACCOUNT_ADDRESS, TEST_PRIVATE_KEY, createSignedMessage, Message} from '@universal-login/commons';
 
-
-export const testPrivateKey = '0x63f01680950dc70f2eb8f373de0c360fcbb89ef437f2f6f2f0a1797979e490a4';
-export const getTestSignedMessage = async (overrides?: Message) => createSignedMessage({from: TEST_ACCOUNT_ADDRESS, to: TEST_ACCOUNT_ADDRESS, ...overrides}, testPrivateKey);
+export const getTestSignedMessage = async (overrides?: Message) => createSignedMessage({from: TEST_ACCOUNT_ADDRESS, to: TEST_ACCOUNT_ADDRESS, ...overrides}, TEST_PRIVATE_KEY);
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/config/message.ts]: 👍

File: universal-login-relayer/test/relayer/routes/wallet.js [hunk 11]
```diff
@@ -70,11 +70,17 @@ describe('E2E: Relayer - WalletContract routes', async () => {
...
         .send(signedMessage);
       expect(status).to.eq(201);
-      expect(await provider.getBalance(otherWallet.address)).to.eq(balanceBefore.add(msg.value));
+      await waitExpect(async () => expect(await provider.getBalance(otherWallet.address)).to.eq(balanceBefore.add(msg.value)));
+      const checkStatusId = async () => {
+        const statusById = await chai.request(relayer.server)
+          .get(`/wallet/execution/status/${body.transaction}`);
+        expect(statusById.body.transactionHash).to.not.be.null;
+      };
+      await waitExpect(() => checkStatusId());
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/relayer/routes/wallet.js]: No need to use ugly waitExpect, instead you can use `stopAndWait()` to make sure queue will be empty before using `expect`.

File: universal-login-relayer/test/relayer/services/MessageHandler.js [hunk 12]
```diff
@@ -30,21 +31,27 @@ describe('Relayer - MessageHandler', async () => {
   it('Error when not enough tokens', async () => {
     const message = {...msg, gasLimit: utils.parseEther('2.0')};
     const signedMessage = await createSignedMessage(message, wallet.privateKey);
-    await expect(messageHandler.handleMessage(signedMessage))
-      .to.be.eventually.rejectedWith('Not enough tokens');
+    const id = await messageHandler.handleMessage(signedMessage);
+    await messageHandler.stopLater();
+    const messageEntry = await messageHandler.getQueueStatus(id);
+    expect(messageEntry.error).to.be.eq('Error: Not enough tokens');
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/relayer/services/MessageHandler.js]: 👍

File: universal-login-relayer/lib/routes/wallet.ts (hunk 7 above)
Comment (xitronix) [reply universal-login-relayer/lib/routes/wallet.ts]: Needed only two fields and wanted to change hash for transactionHash until we change two stores to one

File: universal-login-relayer/lib/services/messages/MessageQueueStore.ts (hunk 8 above)
Comment (xitronix) [reply universal-login-relayer/lib/services/messages/MessageQueueStore.ts]: ✔️

File: universal-login-sdk/lib/sdk.ts [hunk 13]
```diff
@@ -127,18 +127,34 @@ class UniversalLoginSDK {
     return this.relayerApi.getConfig();
   }
 
+  async waitForStatus(messageId: string, timeout: number = 1000, tick: number = 20) {
+    let elapsed = 0;
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: let stopTime = now() + timeout();

File: universal-login-sdk/lib/sdk.ts [hunk 14]
```diff
@@ -127,18 +127,34 @@ class UniversalLoginSDK {
...
+    let status;
+    do {
+      if (elapsed > timeout) {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: if (now() > stopTime())

File: universal-login-sdk/lib/sdk.ts [hunk 15]
```diff
@@ -127,18 +127,34 @@ class UniversalLoginSDK {
...
+        throw Error('Timeout');
+      }
+      await sleep(tick);
+      elapsed += tick;
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: unnecessary

File: universal-login-sdk/lib/sdk.ts [hunk 16]
```diff
@@ -127,18 +127,34 @@ class UniversalLoginSDK {
     return this.relayerApi.getConfig();
   }
 
+  async waitForStatus(messageId: string, timeout: number = 1000, tick: number = 20) {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: `timeout: number = 5000, tick: number = 1000`

File: universal-login-relayer/lib/services/messages/PendingMessages.ts (hunk 9 above)
Comment (xitronix) [reply universal-login-relayer/lib/services/messages/PendingMessages.ts]: Message doesn't have field 'id' and shouldn't have. Now we need 'id', but in future single store for Messages will have only messageHash without id. It was simpler change i guess.

File: universal-login-sdk/lib/sdk.ts (hunk 16 above)
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: This function stinks as hell (too long, mixed abstraction layers), her I refactored it for you:

Util function:
```ts
async function retry(callback, timeout: number = 5000, tick: number = 1000) {
  let stopTime = now() + timeout;
  let result;
  do {
    ensure(stopTime <= now(), new Timeout)
    await sleep(tick);    
    result = callback;
  } while (!result);
  return result;
}
```

RelayerApi function:
```ts
async getMessageStatusIfMined() {
  const status =  await this.relayerApi.getStatusById(messageId);
  ensure(!status.error, new APIError(status))
  return status.transactionHash ? status : null; 
}
```

Sdk function:
```ts
async waitForStatus(messageId: string, timeout: number = 5000, tick: number = 1000) {
  return retry(getMessageStatusIfMined);
}
```

Add some nice error hierarchy in newly created `errors.ts`:
```ts
class SDKError extends Error {

}

class APIError extends SDKError {
  constructor(response) {
    super(response.error);
  }
}

class Timeout extends SDKError {

}
```

File: universal-login-sdk/lib/sdk.ts [hunk 17]
```diff
@@ -127,18 +127,34 @@ class UniversalLoginSDK {
...
       nonce: message.nonce || parseInt(await this.getNonce(message.from!, privateKey), 10),
     } as MessageWithFrom;
     const signedMessage = await createSignedMessage(unsignedMessage, privateKey);
-
-    const result = await this.relayerApi.execute(signedMessage);
-
-    if (result.transaction.hash) {
-      const transactionHash = result.transaction.hash;
+    const result = await this.relayerApi.execute(stringifySignedMessageFields(signedMessage));
+    if (typeof result.transaction === 'number') {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: Why would `result.transaction` type be number?

File: universal-login-sdk/lib/sdk.ts [hunk 18]
```diff
@@ -127,18 +127,34 @@ class UniversalLoginSDK {
...
+      const status = await this.waitForStatus(result.transaction);
+      const {transactionHash} = status;
       await waitToBeMined(this.provider, transactionHash);
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: Isn't it already mined after `waitForStatus`?

File: universal-login-sdk/test/sdk.js [hunk 19]
```diff
@@ -67,7 +67,8 @@ describe('SDK - integration', async () => {
     describe('Execute signed message', async () => {
       it('Should execute signed message', async () => {
         const expectedBalance = (await otherWallet.getBalance()).add(utils.parseEther('0.5'));
-        await sdk.execute({...message, to: otherWallet.address}, privateKey);
+        const transactionHash = await sdk.execute({...message, to: otherWallet.address}, privateKey);
+        expect(transactionHash).to.match(/^[0x|0-9|a-f|A-F]{66}/);
```
Comment (marekkirejczyk) [inline universal-login-sdk/test/sdk.js]: 👍

File: universal-login-wallet/test/integration/services/LoginServices.ts [hunk 20]
```diff
@@ -67,8 +67,7 @@ describe('Login', () => {
         privateKey,
         {gasToken: ETHER_NATIVE_TOKEN.address, gasPrice: DEFAULT_GAS_PRICE, gasLimit: DEFAULT_GAS_LIMIT},
         MANAGEMENT_KEY);
-      await waitUntil(() => !!callback.firstCall);
-      expect(callback).to.have.been.calledOnce;
+      await waitExpect(() => expect(!!callback.firstCall).to.be.true);
```
Comment (marekkirejczyk) [inline universal-login-wallet/test/integration/services/LoginServices.ts]: 👍

File: universal-login-relayer/lib/services/messages/MessageQueueStore.ts [hunk 21]
```diff
@@ -39,11 +38,15 @@ export default class MessageQueueStore implements IMessageQueueStore {
...
+    await this.update(messageId, {hash: transactionHash});
   }
 
-  async markAsError (messageId: string, error: string) {
-    await this.remove(messageId, {error});
+  async markAsError (messageId: string, inputError: string) {
+    let error: string = inputError;
+    if (error.length > 255) {
+      error = error.substr(0, 255);
+    }
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/messages/MessageQueueStore.ts]: varchar?

File: universal-login-relayer/lib/services/messages/PendingMessages.ts [hunk 22]
```diff
@@ -33,9 +32,8 @@ export default class PendingMessages {
   private async onReadyToExecute(messageHash: string, message: SignedMessage) {
     const finalMessage = await this.getMessageWithSignatures(message, messageHash);
     await this.ensureCorrectExecution(messageHash);
-    const transaction: providers.TransactionResponse = await this.doExecute(finalMessage);
-    await this.confirmExecution(messageHash, transaction.hash!);
-    return transaction;
+    const id = await this.messageQueue.add(finalMessage);
+    return id && id[0];
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/services/messages/PendingMessages.ts]: return id[0];

Comment (xitronix) [reply universal-login-relayer/lib/services/messages/PendingMessages.ts]: ✔️

File: universal-login-sdk/lib/sdk.ts (hunk 18 above)
Comment (xitronix) [reply universal-login-sdk/lib/sdk.ts]: 👍 ✔️

File: universal-login-sdk/lib/sdk.ts (hunk 17 above)
Comment (xitronix) [reply universal-login-sdk/lib/sdk.ts]: Changed ✔️

File: universal-login-relayer/lib/services/messages/PendingMessages.ts (hunk 9 above)
Comment (xitronix) [reply universal-login-relayer/lib/services/messages/PendingMessages.ts]: Changed ✔️

File: universal-login-sdk/lib/sdk.ts (hunk 16 above)
Comment (xitronix) [reply universal-login-sdk/lib/sdk.ts]: #nextPR

File: universal-login-relayer/lib/services/messages/MessageExecutor.ts (hunk 4 above)
Comment (xitronix) [reply universal-login-relayer/lib/services/messages/MessageExecutor.ts]: ✔️

File: universal-login-relayer/lib/services/messages/MessageQueueStore.ts (hunk 21 above)
Comment (marekkirejczyk) [reply universal-login-relayer/lib/services/messages/MessageQueueStore.ts]: text

File: universal-login-relayer/test/helpers/MessageQueueMemoryStore.ts [hunk 23]
```diff
@@ -18,7 +18,7 @@ export default class MessageQueueMemoryStore implements IMessageQueueStore {
       hash: 'hash',
       error: undefined
     });
-    return this.counter.toString();
+    return [this.counter.toString()];
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/helpers/MessageQueueMemoryStore.ts]: no []

File: universal-login-relayer/lib/routes/wallet.ts (hunk 6 above)
Comment (marekkirejczyk) [reply universal-login-relayer/lib/routes/wallet.ts]: Ticket created

**Pull request merged**